File: FBControlCoreTests/Tests/Unit/AXTraitsTest.m
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import <XCTest/XCTest.h>

#import <AXRuntime/AXTraits.h>
#import <FBControlCore/FBControlCore.h>

@interface AXTraitsTest : XCTestCase
@end

@implementation AXTraitsTest

- (void)testMappingNames
{
  NSDictionary<NSNumber *, NSString *> *mapping = AXTraitToNameMap();
  XCTAssertEqualObjects(mapping[@(AXTraitLink)], @"Link");
  XCTAssertEqualObjects(mapping[@(AXTraitButton)], @"Button");
}

- (void)testEmptyTraitExtraction
{
  XCTAssertEqualObjects(AXExtractTraits(AXTraitNone), [NSSet setWithObject:@"None"]);
}

- (void)testSingleTraitExtraction
{
  XCTAssertEqualObjects(AXExtractTraits(AXTraitButton), [NSSet setWithObject:@"Button"]);
}

- (void)testUnknownTraitExtraction
{
  XCTAssertEqualObjects(AXExtractTraits((uint64)1 << 63), [NSSet setWithObject:@"Unknown"]);
}

- (void)testCombinedTraitExtraction
{
  NSSet *expectedSet = [NSSet setWithObjects:@"Button", @"Selected", nil];
  XCTAssertEqualObjects(AXExtractTraits(AXTraitButton | AXTraitSelected), expectedSet);
}

- (void)testCombinedTraitExtractionWithUnknownTrait
{
  NSSet *expectedSet = [NSSet setWithObjects:@"Button", @"Unknown", nil];
  XCTAssertEqualObjects(AXExtractTraits(AXTraitButton | (uint64)1 << 63), expectedSet);
}

@end

File: FBControlCoreTests/Tests/Unit/AXTraitsTest.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

// Disabled during swift-format 6.3 rollout, feel free to remove:
// swift-format-ignore-file: OrderedImports

import AXRuntime
import XCTest

@testable import FBControlCore

final class AXTraitsTest: XCTestCase {
  func testMappingNames() {
    let mapping = AXTraitToNameMap()
    XCTAssertEqual(mapping[NSNumber(value: AXTraits.link.rawValue)], "Link")
    XCTAssertEqual(mapping[NSNumber(value: AXTraits.button.rawValue)], "Button")
  }

  func testEmptyTraitExtraction() {
    XCTAssertEqual(AXExtractTraits(0), Set(["None"]))
  }

  func testSingleTraitExtraction() {
    XCTAssertEqual(AXExtractTraits(AXTraits.button.rawValue), Set(["Button"]))
  }

  func testUnknownTraitExtraction() {
    XCTAssertEqual(AXExtractTraits(UInt64(1) << 63), Set(["Unknown"]))
  }

  func testCombinedTraitExtraction() {
    let expectedSet: Set<String> = Set(["Button", "Selected"])
    XCTAssertEqual(AXExtractTraits(AXTraits.button.rawValue | AXTraits.selected.rawValue), expectedSet)
  }

  func testCombinedTraitExtractionWithUnknownTrait() {
    let expectedSet: Set<String> = Set(["Button", "Unknown"])
    XCTAssertEqual(AXExtractTraits(AXTraits.button.rawValue | (UInt64(1) << 63)), expectedSet)
  }
}
